model_discretization_api.py

```diff
@@ -0,0 +1,209 @@
+'''
+A framework for efficient model image rasterization in astropy. 
+
+This is a GSoC 2015 project and is therefore on a relatively short time frame. 
+We focus first on developing a robust solution to a simple but common use case that 
+is extensible to future implementation of more complex models and solutions. 
+
+Use case: Generate a sky model image of discrete sources, e.g. something like 
+photutils.datasets.make_100gaussians_image(), but faster and convolving a source model
+with an instrument PSF. 
+
+The primary challenge- and motivation- is maximizing speed while maintainging precision. A 
+straightforward approach is to use oversampling within a bounded area around the 
+source containing the majority of its flux (e.g. 5-sigma) for discretization and integration. 
+
+The key steps are:
+1. Model generation
+	- Convolution of a source model with an instrument PSF model
+	- Fitting the convolved model to image data if required
+2. Model discretization/Image rasterization
+	- Integration of the model over pixel areas
+	- Reprojection if required
+
+The pieces necessary are largely in existence. Relevant packages 
+include astropy.modeling, astropy.convolution, astropy.nddata, sherpa, galfit, reproject,
+... others.
+
+Outlined here is a rough draft of the proposed API for community discussion. 
+
+'''
+
+######################
+# ConvolvedModel API #
+######################
+
+from astropy.modeling import ConvolvedModel
+from astropy.modeling.models import Box2D, Lorentz1D, Gaussian2D
+from astropy.convolution import Gaussian2DKernel, Gaussian1DKernel
+
+# Initiate a ConvolvedModel instance, takes source model and convolution kernel
+# like:
+source = Box2D(1, 0, 0, 1, 1)
+psf = Gaussian2DKernel(1)
+convolved_model = ConvolvedModel(source, psf, mode='oversample', factor=10)
+
+# mode will be used both for convolution and later integration
+# if a different mode is desired for convolution it should be passed in the psf
+
+# support of 1D models
+source = Lorentz1D(1, 0, 0, 1, 1)
+psf = Gaussian1DKernel(1)
+convolved_model = ConvolvedModel(source, psf)
+
+# should there be a Convolved1DModel and Convolved2DModel?
+# should this be an extension of the CompoundModel class?
```

Comment:
Probably not exactly--the `CompoundModel` class is specifically for handling models that are composed out of multiple models that have been combined with arithmetic and/or composition operators.  That said, I can see why you would suggest that here, since you're basically convolving the source model with some convolution kernel, so in that sense you're composing the source model with the convolution operator.

So the way Perry and I talked about doing this is instead of making a `ConvolvedModel` that takes the source as an argument, as proposed here, there might be some `ConvolutionModel` that is instantiated just with the desired kernel and then would run a source model through it like:

```python
convolution = ConvolutionModel(psf)
convolved_model = source | convolution
```

this would require that the `ConvolutionModel` overrides how the pipe operator is implemented for models, since it has a slightly different sense than it does in normal compound models.  In normal compound models, `model_1 | model_2` means "pass the output of model_1 in as the input to model_2".  However, in this case the input to the convolution model is the *entire* model on the LHS (as well as its input, if that makes sense).   This special case could definitely be supported though, I think.  Last we talked about this I said I would whip up a prototype but haven't had a chance.  I can show you more exactly what I mean later if that sounds interesting.

However, if the scheme I'm proposing doesn't work out, something like the above example could be made to work too.  I just like the `source | convolution` scheme since it fits in more nicely overall with the existing syntax for compound models :)